core/modules/map.rs

```diff
@@ -702,155 +700,115 @@ impl ModuleMap {
-    let maybe_module_evaluation = self.pending_mod_evaluate.borrow_mut().take();
-    if maybe_module_evaluation.is_none() {
-      return;
-    }
 
-    let mut module_evaluation = maybe_module_evaluation.unwrap();
+      // Create a ModEvaluate instance and stash it in an external
+      let evaluation = v8::External::new(
+        tc_scope,
+        Box::into_raw(Box::new(ModEvaluate {
+          module_map: self.clone(),
+          sender: Some(sender),
+        })) as _,
+      );
 
-    let promise_global = module_evaluation.promise.clone().unwrap();
-    let promise = promise_global.open(scope);
-    let promise_state = promise.state();
-    match promise_state {
-      v8::PromiseState::Pending => {
-        // NOTE: `poll_event_loop` will decide if
-        // runtime would be woken soon
-        *self.pending_mod_evaluate.borrow_mut() = Some(module_evaluation);
+      fn get_sender(arg: v8::Local<v8::Value>) -> ModEvaluate {
+        let sender = v8::Local::<v8::External>::try_from(arg).unwrap();
+        *unsafe { Box::from_raw(sender.value() as _) }
       }
-      v8::PromiseState::Fulfilled => {
-        scope.perform_microtask_checkpoint();
-        // Receiver end might have been already dropped, ignore the result
-        let _ = module_evaluation.sender.send(Ok(()));
-        module_evaluation.handled_promise_rejections.clear();
-      }
-      v8::PromiseState::Rejected => {
-        let exception = promise.result(scope);
-        scope.perform_microtask_checkpoint();
-
-        // Receiver end might have been already dropped, ignore the result
-        if module_evaluation
-          .handled_promise_rejections
-          .contains(&promise_global)
-        {
-          let _ = module_evaluation.sender.send(Ok(()));
-          module_evaluation.handled_promise_rejections.clear();
-        } else {
-          let _ = module_evaluation
-            .sender
-            .send(exception_to_err_result(scope, exception, false));
+
+      let on_fulfilled = Function::builder(
+        |_scope: &mut v8::HandleScope<'_>,
+         args: v8::FunctionCallbackArguments<'_>,
+         _rv: v8::ReturnValue| {
+          let mut sender = get_sender(args.data());
+          sender.module_map.pending_mod_evaluation.set(false);
+          sender.module_map.module_waker.wake();
+          _ = sender.sender.take().unwrap().send(Ok(()));
+        },
+      )
+      .data(evaluation.into())
+      .build(tc_scope);
+
+      let on_rejected = Function::builder(
+        |scope: &mut v8::HandleScope<'_>,
+         args: v8::FunctionCallbackArguments<'_>,
+         _rv: v8::ReturnValue| {
+          let mut sender = get_sender(args.data());
+          sender.module_map.pending_mod_evaluation.set(false);
+          sender.module_map.module_waker.wake();
+          _ = sender.sender.take().unwrap().send(Ok(()));
+          scope.throw_exception(args.get(0));
+        },
+      )
+      .data(evaluation.into())
+      .build(tc_scope);
```

Comment:
IIRC we discussed to hoist these functions to top-level. Maybe consider still doing that?

Reply:
I tried to do that but the way that the rusty_v8 mapping stuff works makes this extremely difficult to do. I removed as much duplication as I could.

I think we'll need to upstream changes to make these functions more ergonomic as it's generally painful to use them w/closures and I don't think it needs to be.